python/dcr_template.py

```diff
@@ -39,61 +40,132 @@
 
-        @param obsid_range  single, or list of observation IDs in repository to create matched
-                            templates for. Ignored if exposures are supplied directly.
-        @param exposures  optional, list or generator of exposure objects that will
-                          have matched templates created.
-        @param add_noise  If set to true, add Poisson noise to the template based on the variance.
-        @param repository  path to the repository where the exposure data to be matched are stored.
-                           Ignored if exposures are supplied directly.
-        @param output_repository  path to repository directory where templates will be saved.
-        @param kernel_size  [optional] size, in pixels, of the region surrounding each image pixel that DCR
-                            shifts are calculated. Default is to use the same value the model was created with
-        @param instrument  Name of the observatory.
-        @param warp  Flag. Set to true if the exposures have different wcs from the model.
-                     If True, the generated templates will be warped to match the wcs of each exposure.
-        @param verbose  Flag, set to True to print progress messages.
-        @param use_nonnegative  Flag, set to True to use a true non-negative least squares solution [SLOW]
-        @return Returns a generator that builds DCR-matched templates for each exposure.
+        Parameters
+        ----------
+        obsid_range : int, or list of ints, optional
+            Single, or list of observation IDs in repository to create matched
+            templates for. Ignored if exposures are supplied directly.
+        exposures : List or generator of lsst.afw.image.ExposureD objects, optional
+            List or generator of exposure objects that will have matched templates created.
+        add_noise : bool, optional
+            If set to true, add Poisson noise to the template based on the variance.
+        repository : str, optional
+            Path to the repository where the exposure data to be matched are stored.
+            Ignored if exposures are supplied directly.
+        output_repository : str, optional
+            Path to repository directory where templates will be saved.
+            The templates will not be written to disk if `output_repository` is None.
+        instrument : str, optional
+            Name of the observatory.
+        warp : bool, optional
+            Set to true if the exposures have different wcs from the model.
+            If True, the generated templates will be warped to match the wcs of each exposure.
+        verbose : bool, optional
+            Set to True to print progress messages.
+        output_obsid_offset : int, optional
+            Optional offset to add to the output obsids.
+            Use if writing to the same repository as the input to avoid over-writing the input data.
```

Comment:
This was motivated by the limitations of using calexps as templates in image differencing, which seems to require the template calexp be an obsid in the same repository. I agree this is a terrible workaround, so I have created DM-9852 to do this properly.